apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,872 @@
+
+    /* Action related errors */
+    string internal constant ERROR_ACTION_DOES_NOT_EXIST = "AGR_ACTION_DOES_NOT_EXIST";
+    string internal constant ERROR_DISPUTE_DOES_NOT_EXIST = "AGR_DISPUTE_DOES_NOT_EXIST";
+    string internal constant ERROR_CANNOT_CANCEL_ACTION = "AGR_CANNOT_CANCEL_ACTION";
+    string internal constant ERROR_CANNOT_EXECUTE_ACTION = "AGR_CANNOT_EXECUTE_ACTION";
+    string internal constant ERROR_CANNOT_CHALLENGE_ACTION = "AGR_CANNOT_CHALLENGE_ACTION";
+    string internal constant ERROR_CANNOT_SETTLE_ACTION = "AGR_CANNOT_SETTLE_ACTION";
+    string internal constant ERROR_CANNOT_DISPUTE_ACTION = "AGR_CANNOT_DISPUTE_ACTION";
+    string internal constant ERROR_CANNOT_RULE_ACTION = "AGR_CANNOT_RULE_ACTION";
+    string internal constant ERROR_CANNOT_SUBMIT_EVIDENCE = "AGR_CANNOT_SUBMIT_EVIDENCE";
+
+    /* Evidence related errors */
+    string internal constant ERROR_SUBMITTER_FINISHED_EVIDENCE = "AGR_SUBMITTER_FINISHED_EVIDENCE";
+    string internal constant ERROR_CHALLENGER_FINISHED_EVIDENCE = "AGR_CHALLENGER_FINISHED_EVIDENCE";
+
+    /* Arbitrator related errors */
+    string internal constant ERROR_ARBITRATOR_NOT_CONTRACT = "AGR_ARBITRATOR_NOT_CONTRACT";
+    string internal constant ERROR_ARBITRATOR_FEE_RETURN_FAILED = "AGR_ARBITRATOR_FEE_RETURN_FAIL";
+    string internal constant ERROR_ARBITRATOR_FEE_DEPOSIT_FAILED = "AGR_ARBITRATOR_FEE_DEPOSIT_FAIL";
+    string internal constant ERROR_ARBITRATOR_FEE_APPROVAL_FAILED = "AGR_ARBITRATOR_FEE_APPROVAL_FAIL";
+    string internal constant ERROR_ARBITRATOR_FEE_TRANSFER_FAILED = "AGR_ARBITRATOR_FEE_TRANSFER_FAIL";
+
+    /* Collateral token related errors */
+    string internal constant ERROR_COLLATERAL_TOKEN_NOT_CONTRACT = "AGR_COL_TOKEN_NOT_CONTRACT";
+    string internal constant ERROR_COLLATERAL_TOKEN_TRANSFER_FAILED = "AGR_COL_TOKEN_TRANSFER_FAILED";
+
+    // bytes32 public constant STAKE_ROLE = keccak256("STAKE_ROLE");
+    bytes32 public constant STAKE_ROLE = 0xeaea87345c0a5b2ecb49cde771d9ac5bfe2528357e00d43a1e06a12c2779f3ca;
+
+    // bytes32 public constant CHALLENGE_ROLE = keccak256("CHALLENGE_ROLE");
+    bytes32 public constant CHALLENGE_ROLE = 0xef025787d7cd1a96d9014b8dc7b44899b8c1350859fb9e1e05f5a546dd65158d;
+
+    // bytes32 public constant CHANGE_AGREEMENT_ROLE = keccak256("CHANGE_AGREEMENT_ROLE");
+    bytes32 public constant CHANGE_AGREEMENT_ROLE = 0x4af6231bf2561f502301de36b9a7706e940a025496b174607b9d2f58f9840b46;
+
+    event ActionScheduled(uint256 indexed actionId);
+    event ActionChallenged(uint256 indexed actionId);
+    event ActionSettled(uint256 indexed actionId);
+    event ActionDisputed(uint256 indexed actionId);
+    event ActionAccepted(uint256 indexed actionId);
+    event ActionVoided(uint256 indexed actionId);
+    event ActionRejected(uint256 indexed actionId);
+    event ActionCancelled(uint256 indexed actionId);
+    event ActionExecuted(uint256 indexed actionId);
+    event SettingChanged(uint256 indexed settingId);
```

Comment:
I imagine that these are provisional, but we should think about what other data we want to log to make state reducing easier

Reply:
Yup, it was provisional. I added more information now